_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,704 @@
+
+        KafkaProducer<String, String> producer = new KafkaProducer<>(props);
+
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            System.out.println("Shutdown hook triggered. Cleaning up...");
+            producer.close();
+        }));
+
+        int recordNum = 1;
+        while (true) {
+            String recordKey = null;
+            String recordValue;
+            if (recordNum % 10 == 0) {
+                // Send every 10th record with in not processable format
+                recordValue = "Poison message: this record is not in JSON format (recordNumber: " + recordNum + ")";
+            } else if(recordNum % 5 == 0) {
+                // Send every 5th record with key "skip"
+                recordKey = "skip";
+                recordValue = "Record to skip processing";
+            } else {
+                recordValue = "{ \"isJson\": \"true\", \"recordNumber\": " + recordNum + ", \"recordValue\": \"This is test message\" }";
+            }
+
+            ProducerRecord record = new ProducerRecord<>("kafka-queue-topic", recordKey, recordValue);
+            RecordMetadata metadata = (RecordMetadata) producer.send(record).get();
+            System.out.println("Produced record at offset " + metadata.offset() + ": " + record.value());
+            recordNum++;
+            Thread.sleep(Long.valueOf(2000));
+        }
+    }
+}
+```
+
+#### <i>Inspecting __share_group_state topic</i>
+
+Let’s now take a look at the output from `ShareGroupDemoExplicitAck` while running `ShareGroupDemoProducer` at the same time:
+```sh
+Failed to process the record- Offset: 361 Value: Poison message: this record is not in JSON format (recordNumber: 10)
+Processed record with offset 362: {
+  "isJson" : "true",
+  "recordNumber" : 11,
+  "recordValue" : "This is test message"
+}
+Processed record with offset 363: {
+  "isJson" : "true",
+  "recordNumber" : 12,
+  "recordValue" : "This is test message"
+}
+Processed record with offset 364: {
+  "isJson" : "true",
+  "recordNumber" : 13,
+  "recordValue" : "This is test message"
+}
+Processed record with offset 365: {
+  "isJson" : "true",
+  "recordNumber" : 14,
+  "recordValue" : "This is test message"
+}
+Skipping the record - Offset: 366 Key: skip Value: Record to skip processing Delivery count: 1
+Skipping the record - Offset: 366 Key: skip Value: Record to skip processing Delivery count: 2
+Skipping the record - Offset: 366 Key: skip Value: Record to skip processing Delivery count: 3
+Skipping the record - Offset: 366 Key: skip Value: Record to skip processing Delivery count: 4
+Skipping the record - Offset: 366 Key: skip Value: Record to skip processing Delivery count: 5
+Processed record with offset 367: {
+  "isJson" : "true",
+  "recordNumber" : 16,
+  "recordValue" : "This is test message"
+}
+```
+
+The record at offset 361 was rejected because it hit an error mapping the record value into JSON therefore is considered a poison message. 
```

Comment:
```suggestion
The record at offset 361 was rejected because it hit an error mapping the record value into JSON and was considered a poison message. 
```